ethdev/ethdev.go

```diff
@@ -835,6 +837,16 @@ func CountAvail() int {
 	return int(C.rte_eth_dev_count_avail())
 }
 
+// ValidPorts returns list of devices valid in EAL.
+func ValidPorts() (res []Port) {
+	for x := C.rte_eth_find_next(0); x != C.RTE_MAX_ETHPORTS; x = C.rte_eth_find_next(x + 1) {
+		if p := Port(x); p.IsValid() {
+			res = append(res, p)
+		}
+	}
+	return
+}
```

Comment:
I think it's better to wrap in one Cgo call. And moreover, `RTE_ETH_FOREACH_DEV` is the official ideological way to do it. Feel free to polish.
```
static int go_rte_get_ports(uint16_t *ports, int n_ports) {
	uint16_t port;
	int n = 0;

	RTE_ETH_FOREACH_DEV(port) {
		ports[n++] = port;
	}

	return n;
}
```

and then:

```
func ValidPorts() []Port {
	max := CountTotal()
	if max == 0 {
		return nil
	}

	ports := make([]Port, CountTotal())
	return ports[:C.go_rte_get_ports((*C.uint16_t)(unsafe.Pointer(&ports[0])), C.int(max))]
}
```

Reply:
Okay, no problem, I assumed it didn't require good performance.

Reply:
Well, it's more of a broader compatibility issue here rather than performance. Macro is the way to scroll through the ports as it is described in [rte_eth_dev_count_avail()](https://doc.dpdk.org/api/rte__ethdev_8h.html) doc.

Reply:
Fixed